### PR TITLE
Displaying alert if user is viewing an old version of a dataset

### PR DESCRIPTION
**Pull request by RichardHobbsPA**

### CODE REVIEW

**Review by PaulMcCaffertyPA — DISMISSED** [review]

**Review by CMarksPA — APPROVED** [review]

**Review by PaulMcCaffertyPA — DISMISSED** [review]

File: src/pages/course/AddEditCoursePage.js [hunk 1]
```diff
@@ -211,20 +211,31 @@ class AddEditCoursePage extends React.Component {
...
+			}
+			else{
+				this.state.relatedObjects.push({
+					objectId: object.objectId,
+					reason: '',
+					objectType: object.type,
+					user: this.state.userState[0].name,
+					updated: moment().format('DD MMM YYYY')
+				});
+			}
```
Comment (CMarksPA) [inline src/pages/course/AddEditCoursePage.js:238]: @RichardHobbsPA Just a note on setState, React will complain about this and throw and error of incorrect use and goes against their best practice. It is incorrect use of this method, you can read here for more info [React SetState](https://reactjs.org/docs/faq-state.html#what-does-setstate-do)

What I would recommend doing is not to mutate the original state which is important to stop any bugs and unwanted side effects (mutation) from occuring.

Line 219 after function declaration of addToRelatedObjects declare new user variable using desturctuing technique to get the user, save you having to write  _user: this.state.userState[0].name_, 3x in the other blocks
`let {userState: [user, ...rest]} = this.state;`

Line 220 should be something like the following which stops mutation as your writing back into state directly and therefore incorrectly.

// this takes a copy of state relatedObjectIds using spread in the values and stops mutation
`let relatedObjectIds = [...this.state.tempRelatedObjectIds];
let relatedObjects = [...this.state.relatedObjects] `

Line 221 - 238 loop via map as you have done but declare a new let variables and spread the object and alter the values as you need into a new array firstly
`let newRelatedObjects = relatedObjectIds.map((relatedObject) => {
     let newRelatedObject = {
        ...relatedObject,
       objectId: relatedObject.type === 'dataset' ? relatedObject.pid : relatedObject.objectId,
       user: user.name
       updated: moment().format('DD MM YYYY')
     }
    return newRelatedObject;
});`

lastly set the state using setState function in React as below.
`this.setState({relatedObjects: [relatedObjectIds, ....newRelatedObjects]})`

This looks to be the case for the rest of the files too.

File: src/pages/course/AddEditCoursePage.js [hunk 2]
```diff
@@ -234,11 +245,22 @@ class AddEditCoursePage extends React.Component {
...
 	};
 
-	removeObject = (id) => {
-    let updatedRelatedObjects;
+	removeObject = (id, type, datasetid) => {
+
+		let countOfRelatedObjects = this.state.relatedObjects.length;
+		let updatedRelatedObjects;
+		
 		updatedRelatedObjects = this.state.relatedObjects.filter(
```
Comment (CMarksPA) [inline src/pages/course/AddEditCoursePage.js:253]: @RichardHobbsPA again I would try not to mutate and assign in directly.

let updatedRelatedObjects = [...this.state.relatedObjects].filter();`

File: src/pages/paper/AddEditPaperPage.js [hunk 3]
```diff
@@ -257,20 +257,31 @@ class AddEditPaperPage extends React.Component {
...
+			}
+			else{
+				this.state.relatedObjects.push({
+					objectId: object.objectId,
+					reason: '',
+					objectType: object.type,
+					user: this.state.userState[0].name,
+					updated: moment().format('DD MMM YYYY')
+				});
+			}
```
Comment (CMarksPA) [inline src/pages/paper/AddEditPaperPage.js:284]: @RichardHobbsPA this is the same as the previous comments.

File: src/pages/paper/AddEditPaperPage.js [hunk 4]
```diff
@@ -280,14 +291,21 @@ class AddEditPaperPage extends React.Component {
...
-		this.state.relatedObjects = this.state.relatedObjects.filter(
-			(obj) => obj.objectId !== id
-		);
-		this.state.relatedObjects = this.state.relatedObjects.filter(
-			(obj) => obj.objectId !== id.toString()
+	removeObject = (id, type, datasetid) => {
+
+		let countOfRelatedObjects = this.state.relatedObjects.length;
+		let newRelatedObjects = this.state.relatedObjects.filter(
+			(obj) => (obj.objectId !== id && obj.objectId !== id.toString())
```
Comment (CMarksPA) [inline src/pages/paper/AddEditPaperPage.js:298]: @RichardHobbsPA you might want to update this area too to stay consistent with the other changes.

`let updatedRelatedObjects = [...this.state.relatedObjects].filter();`

File: src/pages/paper/AddEditPaperPage.js [hunk 5]
```diff
@@ -280,14 +291,21 @@ class AddEditPaperPage extends React.Component {
...
 		);
-		this.setState({ relatedObjects: this.state.relatedObjects });
+		
+		//if an item was not removed try removing by datasetid for retro linkages 
+		if((countOfRelatedObjects <= newRelatedObjects.length) && type === 'dataset'){
+			
+			newRelatedObjects = this.state.relatedObjects.filter(
+				(obj) => (obj.objectId !== datasetid && obj.objectId !== datasetid.toString())
+			);
```
Comment (CMarksPA) [inline src/pages/paper/AddEditPaperPage.js:306]: @RichardHobbsPA 

Line 304 - 306
newRelatedObjects = [...this.state.relatedObjects].filter(
				(obj) => (obj.objectId !== datasetid && obj.objectId !== datasetid.toString())
			);

File: src/pages/project/AddEditProjectPage.js [hunk 6]
```diff
@@ -155,20 +155,24 @@ class AddEditProjectPage extends React.Component {
...
 
     addToRelatedObjects = () => {
         this.state.tempRelatedObjectIds.map((object) => {
-            this.state.relatedObjects.push({'objectId':object.objectId, 'reason':'', 'objectType':object.type, 'user':this.state.userState[0].name, 'updated':moment().format("DD MMM YYYY")})
+            if(object.type === 'dataset'){
+                this.state.relatedObjects.push({'objectId':object.pid, 'reason':'', 'objectType':object.type, 'user':this.state.userState[0].name, 'updated':moment().format("DD MMM YYYY")})
+            }
+            else{
+                this.state.relatedObjects.push({'objectId':object.objectId, 'reason':'', 'objectType':object.type, 'user':this.state.userState[0].name, 'updated':moment().format("DD MMM YYYY")})
+            }
```
Comment (CMarksPA) [inline src/pages/project/AddEditProjectPage.js:175]: @RichardHobbsPA this is the same issue with state and needs to be this.setState

**Review by CMarksPA — COMMENTED** [review]
Richard I have made a few comments, it might be good to change the state updating factor as it will cause an issue at somepoint.

**Review by CiaraWardPA — APPROVED** [review]

**Pull request merged**